canvas/async_enroll_gatherer.py

```diff
@@ -0,0 +1,210 @@
+                    course_enrollment_dict = self.course_enrollments[course_id]
+                    enroll_page_info = course_enrollment_dict['page_info']
+                    params['variables']['enrollmentPageCursor'] = enroll_page_info['endCursor']
+
+                logger.debug(params['variables'])
+                response = session.post(self.complete_url, json=params)
+                responses.append(response)
+
+            for completed_response in as_completed(responses):
+                self.parse_enrollment_response(completed_response)
+
+                # Log process status
+                logger.info(f'# started courses: {len(self.course_enrollments)}')
+                logger.info(f'# completed courses: {len(self.get_complete_course_ids())}')
+
+    def generate_output(self) -> Tuple[pd.DataFrame, ...]:
+        logger.debug('generate_output')
+        enrollment_records = []
+        user_records = []
+        section_records = []
+
+        for course_id in self.course_enrollments.keys():
+            enrollment_dicts = self.course_enrollments[course_id]['enrollments']
+            for enrollment_dict in enrollment_dicts:
+                enrollment_record, user_record, section_record = unnest_enrollment(enrollment_dict)
+                enrollment_records.append(enrollment_record)
+                user_records.append(user_record)
+                section_records.append(section_record)
+
+        # Seems like we shouldn't have to drop duplicates, but once one duplicate broke the process
+        enrollment_df = pd.DataFrame(enrollment_records)
+        enrollment_count = len(enrollment_df)
+        enrollment_df = enrollment_df.drop_duplicates()
+        logger.info(f'{len(enrollment_df) - enrollment_count} enrollment records were dropped')
+
+        user_df = pd.DataFrame(user_records).drop_duplicates()
+        section_df = pd.DataFrame(section_records).drop_duplicates()
+        return (enrollment_df, user_df, section_df)
+
+    def gather(self) -> None:
+        logger.info('** AsyncEnrollGatherer')
+        logger.info('Gathering enrollment data for courses asynchronously with GraphQL')
+
+        more_to_gather = True
+
+        while more_to_gather:
+            complete_course_ids = self.get_complete_course_ids()
+            course_ids_to_process = sorted(self.get_incomplete_course_ids())
+
+            if len(course_ids_to_process) == 0:
+                more_to_gather = False
+            else:
+                unstarted_course_ids = [course_id for course_id in course_ids_to_process if course_id not in self.course_enrollments.keys()]
+                # This condition will stop the gather process if all the remaining course_ids have not been started.
+                # Several request attempts will have been made using this course_id before reaching this state.
+                if len(complete_course_ids) > 0 and unstarted_course_ids == course_ids_to_process:
```

Comment:
I guess you are checking after the initial round processing if the unfinished list doesn't change the exit out?

Reply:
Kind of. I'm checking whether at least one course is complete and whether the remaining courses to process are unstarted, i.e. they haven't successfully processed any pages. Until the end, the course IDs should include started but unfinished courses.